File: backend/app/modules/litigation_approvals_compliance/overview.py
"""Overview summary derived from the Litigation, Approvals & Compliance draft."""

from __future__ import annotations

from typing import Any

from app.modules.litigation_approvals_compliance.assessment import assess_litigation_approvals_compliance
from app.modules.litigation_approvals_compliance.compute import compute_litigation_approvals_compliance_model
from app.modules.litigation_approvals_compliance.constants import SECTION_LABELS
from app.modules.litigation_approvals_compliance.progress import calculate_litigation_approvals_compliance_progress


def build_overview_summary(
    payload: dict[str, Any],
    linked_references: dict[str, Any] | None = None,
) -> dict[str, Any]:
    linked = linked_references or {}
    progress = calculate_litigation_approvals_compliance_progress(payload)
    model = compute_litigation_approvals_compliance_model(payload, linked)
    assessment = assess_litigation_approvals_compliance(payload, linked)
    primary_exposure = (model.get("exposureByCurrency") or [{}])[0].get("totalExposure", "")

    sections_in_progress = sum(
        1 for status in (progress.get("sections") or {}).values() if status == "in_progress"
    )

    incomplete_sections = [
        (section_id, status)
        for section_id, status in (progress.get("sections") or {}).items()
        if status != "complete"
    ]

    recommended_next_actions = [
        {
            "sectionId": section_id,
            "label": f"Continue with {SECTION_LABELS.get(section_id, section_id)}",
        }
        for section_id, _ in incomplete_sections[:4]
    ]

    counts = assessment.get("counts") or {}
    assessment_concerns = (
        counts.get("potentialConcern", 0)
        + counts.get("materialityReviewRequired", 0)
        + counts.get("approvalRenewalReviewRequired", 0)
        + counts.get("complianceReviewRequired", 0)
        + counts.get("pendingLegalReview", 0)
        + counts.get("financialReconciliationPending", 0)
    )
    pending_professional_review_items = counts.get("pendingProfessionalConfirmation", 0) + counts.get(
        "pendingBoardDetermination", 0
    )

    expiry_windows = model.get("approvalExpiryWindows") or {}
    compliance_counts = model.get("complianceCounts") or {}
    creditor_totals = model.get("creditorTotals") or {}
    reconciliation = model.get("reconciliation") or {}
    section7 = payload.get("materialCreditorsPenaltiesAndMaterialDevelopments") or {}

    return {
        "sectionStatuses": progress.get("sections") or {},
        "sectionsComplete": progress.get("sectionsComplete", 0),
        "sectionsInProgress": sections_in_progress,
        "totalSections": progress.get("totalSections", 0),
        "overallStatus": progress.get("overallStatus", "not_started"),
        "legalDdAsOfDate": model.get("legalDdAsOfDate", ""),
        "matterCount": model.get("matterCount", 0),
        "criminalMatterCount": model.get("criminalMatterCount", 0),
        "taxMatterCount": model.get("taxMatterCount", 0),
        "pendingOutcomeCount": model.get("pendingOutcomeCount", 0),
        "primaryExposure": primary_exposure,
        "taxAggregateDemand": (model.get("taxAggregates") or {}).get("totalDemand", ""),
        "approvalCount": model.get("approvalCount", 0),
        "expiredApprovalCount": model.get("expiredApprovalCount", 0),
        "renewalPendingCount": model.get("renewalPendingCount", 0),
        "approvalsExpiringWithin30Days": len(expiry_windows.get("within30Days") or []),
        "approvalsExpiringWithin90Days": len(expiry_windows.get("within90Days") or []),
        "complianceIssueCount": compliance_counts.get("complianceIssueCount", 0),
        "delayedStatutoryDues": compliance_counts.get("delayedStatutoryDues", 0),
        "approvalConditionsOutstanding": compliance_counts.get("approvalConditionsOutstanding", 0),
        "materialCreditorCount": creditor_totals.get("materialCreditorCount", 0),
        "msmeCreditorCount": creditor_totals.get("msmeCreditorCount", 0),
        "creditorAggregateOutstanding": creditor_totals.get("aggregateOutstanding", ""),
        "materialDevelopmentCount": len(section7.get("materialDevelopments") or []),
        "remediationOpenCount": model.get("remediationOpenCount", 0),
        "financialsReconciliationStatus": (reconciliation.get("financials") or {}).get("status", ""),
        "groupEntitiesReconciliationStatus": (reconciliation.get("groupEntities") or {}).get(
            "status", ""
        ),
        "managementGovernanceReconciliationStatus": (
            reconciliation.get("managementGovernance") or {}
        ).get("status", ""),
        "bacReconciliationStatus": (reconciliation.get("bac") or {}).get("status", ""),
        "businessOperationsReconciliationStatus": (reconciliation.get("businessOperations") or {}).get(
            "status", ""
        ),
        "objectsReconciliationStatus": (reconciliation.get("objectsOfIssue") or {}).get("status", ""),
        "ipoSetupReconciliationStatus": (reconciliation.get("ipoSetup") or {}).get("status", ""),
        "assessmentConcerns": assessment_concerns,
        "pendingProfessionalReviewItems": pending_professional_review_items,
        "assessmentResult": assessment.get("result", ""),
        "assessmentResultLabel": assessment.get("resultLabel", ""),
        "assessmentSummary": assessment.get("summary", ""),
        "recommendedNextActions": recommended_next_actions,
    }

File: frontend/lib/litigation-approvals-compliance/overview.ts
/**
 * Overview summary derived from the in-memory Litigation, Approvals & Compliance draft (LAC1).
 */

import {
  assessLitigationApprovalsCompliance,
  type LacAssessmentResponse,
} from '@/lib/litigation-approvals-compliance/assessment';
import {
  computeLitigationApprovalsComplianceModel,
  type LitigationApprovalsComplianceModel,
} from '@/lib/litigation-approvals-compliance/compute';
import { LAC_SECTION_LABELS } from '@/lib/litigation-approvals-compliance/options';
import { calculateLitigationApprovalsComplianceProgress } from '@/lib/litigation-approvals-compliance/progress';
import {
  createEmptyLinkedWorkstreamReferences,
  type LacProgress,
  type LinkedWorkstreamReferences,
} from '@/lib/litigation-approvals-compliance/types';
import type {
  LitigationApprovalsCompliancePayload,
  LitigationApprovalsComplianceSectionId,
} from '@/lib/schemas/litigation-approvals-compliance';

export type LitigationApprovalsComplianceOverviewSummary = {
  sectionStatuses: LacProgress['sections'];
  sectionsComplete: number;
  sectionsInProgress: number;
  totalSections: number;
  overallStatus: LacProgress['overallStatus'];
  legalDdAsOfDate: string;
  matterCount: number;
  criminalMatterCount: number;
  taxMatterCount: number;
  pendingOutcomeCount: number;
  primaryExposure: string;
  taxAggregateDemand: string;
  approvalCount: number;
  expiredApprovalCount: number;
  renewalPendingCount: number;
  approvalsExpiringWithin30Days: number;
  approvalsExpiringWithin90Days: number;
  complianceIssueCount: number;
  delayedStatutoryDues: number;
  approvalConditionsOutstanding: number;
  materialCreditorCount: number;
  msmeCreditorCount: number;
  creditorAggregateOutstanding: string;
  materialDevelopmentCount: number;
  remediationOpenCount: number;
  financialsReconciliationStatus: string;
  groupEntitiesReconciliationStatus: string;
  managementGovernanceReconciliationStatus: string;
  bacReconciliationStatus: string;
  businessOperationsReconciliationStatus: string;
  objectsReconciliationStatus: string;
  ipoSetupReconciliationStatus: string;
  assessmentConcerns: number;
  pendingProfessionalReviewItems: number;
  assessmentResult: LacAssessmentResponse['result'];
  assessmentResultLabel: string;
  assessmentSummary: string;
  recommendedNextActions: Array<{ sectionId: LitigationApprovalsComplianceSectionId; label: string }>;
};

export function buildOverviewSummary(
  payload: LitigationApprovalsCompliancePayload,
  linkedReferences: LinkedWorkstreamReferences = createEmptyLinkedWorkstreamReferences(),
): LitigationApprovalsComplianceOverviewSummary {
  const progress = calculateLitigationApprovalsComplianceProgress(payload);
  const model = computeLitigationApprovalsComplianceModel(payload, linkedReferences);
  const assessment = assessLitigationApprovalsCompliance(payload, linkedReferences);
  const primaryExposure = model.exposureByCurrency[0]?.totalExposure ?? '';

  const sectionsInProgress = Object.values(progress.sections).filter(
    (status) => status === 'in_progress',
  ).length;

  const incompleteSections = (
    Object.entries(progress.sections) as Array<
      [
        LitigationApprovalsComplianceSectionId,
        LacProgress['sections'][LitigationApprovalsComplianceSectionId],
      ]
    >
  ).filter(([, status]) => status !== 'complete');

  const recommendedNextActions = incompleteSections.slice(0, 4).map(([sectionId]) => ({
    sectionId,
    label: `Continue with ${LAC_SECTION_LABELS[sectionId]}`,
  }));

  const assessmentConcerns =
    assessment.counts.potentialConcern +
    assessment.counts.materialityReviewRequired +
    assessment.counts.approvalRenewalReviewRequired +
    assessment.counts.complianceReviewRequired +
    assessment.counts.pendingLegalReview +
    assessment.counts.financialReconciliationPending;

  const pendingProfessionalReviewItems =
    assessment.counts.pendingProfessionalConfirmation +
    assessment.counts.pendingBoardDetermination;

  return {
    sectionStatuses: progress.sections,
    sectionsComplete: progress.sectionsComplete,
    sectionsInProgress,
    totalSections: progress.totalSections,
    overallStatus: progress.overallStatus,
    legalDdAsOfDate: model.legalDdAsOfDate,
    matterCount: model.matterCount,
    criminalMatterCount: model.criminalMatterCount,
    taxMatterCount: model.taxMatterCount,
    pendingOutcomeCount: model.pendingOutcomeCount,
    primaryExposure,
    taxAggregateDemand: model.taxAggregates.totalDemand,
    approvalCount: model.approvalCount,
    expiredApprovalCount: model.expiredApprovalCount,
    renewalPendingCount: model.renewalPendingCount,
    approvalsExpiringWithin30Days: model.approvalExpiryWindows.within30Days.length,
    approvalsExpiringWithin90Days: model.approvalExpiryWindows.within90Days.length,
    complianceIssueCount: model.complianceCounts.complianceIssueCount,
    delayedStatutoryDues: model.complianceCounts.delayedStatutoryDues,
    approvalConditionsOutstanding: model.complianceCounts.approvalConditionsOutstanding,
    materialCreditorCount: model.creditorTotals.materialCreditorCount,
    msmeCreditorCount: model.creditorTotals.msmeCreditorCount,
    creditorAggregateOutstanding: model.creditorTotals.aggregateOutstanding,
    materialDevelopmentCount:
      payload.materialCreditorsPenaltiesAndMaterialDevelopments.materialDevelopments.length,
    remediationOpenCount: model.remediationOpenCount,
    financialsReconciliationStatus: model.reconciliation.financials.status,
    groupEntitiesReconciliationStatus: model.reconciliation.groupEntities.status,
    managementGovernanceReconciliationStatus: model.reconciliation.managementGovernance.status,
    bacReconciliationStatus: model.reconciliation.bac.status,
    businessOperationsReconciliationStatus: model.reconciliation.businessOperations.status,
    objectsReconciliationStatus: model.reconciliation.objectsOfIssue.status,
    ipoSetupReconciliationStatus: model.reconciliation.ipoSetup.status,
    assessmentConcerns,
    pendingProfessionalReviewItems,
    assessmentResult: assessment.result,
    assessmentResultLabel: assessment.resultLabel,
    assessmentSummary: assessment.summary,
    recommendedNextActions,
  };
}

export type { LitigationApprovalsComplianceModel };
